stream.go

```diff
@@ -0,0 +1,769 @@
+package http2
+
+import (
+	"container/list"
+	"fmt"
+)
+
+const (
+	defaultWeight    = 15
+	defaulMaxDepTree = 100
+	maxStreamWeight  = 255
+)
+
+var defaultPriority = PriorityParam{
+	Weight:    defaultWeight,
+	StreamDep: 0,
+}
+
+type streamDepState int
+
+const (
+	depStateIdle     streamDepState = iota // init state without data
+	depStateReady                          // ready to send but blocked by dependency tree
+	depStateTop                            // can send data
+	depStatFlowDefer                       // blocked by stream flow control
+)
+
+var depStateName = [...]string{
+	depStateIdle:     "Idle",
+	depStateReady:    "Ready",
+	depStateTop:      "Top",
+	depStatFlowDefer: "FlowDefer",
+}
+
+func (ds streamDepState) String() string {
+	return depStateName[ds]
+}
+
+// roots is a list of connection's root streams.
+type roots struct {
+	*list.List       // doubly linked list of all root streams
+	n          int32 // number of streams in the dependency tree
+}
+
+func newRoots() *roots {
+	return &roots{list.New(), 0}
+}
+
+// add stream and increment n.
+func (r *roots) add(st *stream) {
+	st.rootElem = r.PushFront(st)
+	r.n += 1
+}
+
+func (r *roots) move(st *stream) {
+	st.rootElem = r.PushFront(st)
+}
+
+// remove removes stream but doesn't decrement n.
+func (r *roots) remove(st *stream) {
+	r.Remove(st.rootElem)
+}
+
+// removeAll removes all streams but doesn't flush n.
+func (r *roots) removeAll() {
+	r.Init()
+}
+
+// stream represents a stream. This is the minimal metadata needed by
+// the serve goroutine. Most of the actual stream state is owned by
+// the http.Handler's goroutine in the responseWriter. Because the
+// responseWriter's responseWriterState is recycled at the end of a
+// handler, this struct intentionally has no pointer to the
+// *responseWriter{,State} itself, as the Handler ending nils out the
+// responseWriter's state field.
+type stream struct {
```

Comment:
Some of this data could be an embedded struct using a `copy(from)` method instead. Since they are not exported, this seems safe and can be changed later easily. But it certainly makes it easier to understand ownership and lifetime.